File: internal/record/devtools.go
package record

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// devtoolsFile is the log journal of one recording.
const devtoolsFile = "devtools.jsonl"

// DefaultMaxLog caps the log journal of one recording.
//
// A broken page can report the same error thousands of times a minute, and a
// busy application makes thousands of requests. The frames are the recording,
// so the log must never be the reason a disk fills up.
const DefaultMaxLog int64 = 20 << 20

// NoMaxLog lifts the cap. Limits.MaxLog is zero when nobody chose, and zero
// means the default, so "no limit" needs a value of its own.
const NoMaxLog int64 = -1

// The kinds a LogEvent can be.
const (
	LogConsole = "console" // a console.log, warn, error, info or debug call
	LogError   = "error"   // an uncaught exception
	LogIssue   = "issue"   // a browser log entry: CORS, CSP, a failed subresource
	LogReq     = "req"     // a request left the browser
	LogRes     = "res"     // it came back
	LogNetFail = "netfail" // it failed, was blocked, or was cancelled
	LogTap     = "tap"     // the log moved to another tab
	LogDrop    = "drop"    // the rate cap refused some lines
)

// LogEvent is one thing the page reported.
//
// It is a plain struct on purpose. The CDP layer builds it and this package
// stores it, so a recording never depends on a browser protocol, and a reader
// years from now needs nothing but the field names.
//
// A LogEvent never carries a request body, a response body, or a header. A
// login post holds the password in its body and the session in its headers,
// and the recorder already refuses to write what somebody typed.
type LogEvent struct {
	// AtMs is the position on the recording clock. The recorder fills it in.
	AtMs int64 `json:"atMs"`
	// TS is the wall clock in unix milliseconds, as the tap saw it. A
	// recording can start at any point in a session, so the tap cannot know
	// AtMs and the recorder cannot know when the event really happened.
	TS int64  `json:"ts"`
	T  string `json:"t"`

	Level string `json:"level,omitempty"` // debug, info, log, warning, error
	Text  string `json:"text,omitempty"`
	Stack string `json:"stack,omitempty"`

	ReqID  string `json:"reqId,omitempty"` // ties a res or a netfail to its req
	Method string `json:"method,omitempty"`
	URL    string `json:"url,omitempty"`
	Status int    `json:"status,omitempty"`
	Kind   string `json:"kind,omitempty"` // document, xhr, fetch, script, image
	Bytes  int64  `json:"bytes,omitempty"`
	DurMs  int64  `json:"durMs,omitempty"`

	// Count says how many identical lines this one stands for. A rate cap drop
	// reports what it refused, and a repeated failure reports what it
	// collapsed. Zero and one both mean one.
	Count    int    `json:"count,omitempty"`
	TargetID string `json:"targetId,omitempty"`
}

// DevTools is what the manifest says about the log journal.
//
// It is present on every recording made since the log existed, even when the
// page said nothing at all. A quiet page and a recording from before the
// feature are different answers, and only a block with a zero in it can give
// the first one.
type DevTools struct {
	Lines   int   `json:"lines"`
	Bytes   int64 `json:"bytes"`
	Dropped int   `json:"dropped"`
	// Errors counts the failures the page reported, not the marks they made.
	// Fifty repeats of one error collapse into one mark, and a list that says
	// "1 error" about that session is telling the wrong story.
	Errors int `json:"errors"`
	// What the capture was allowed to keep. A reader has to be able to tell
	// "this recording kept no headers" from "this recording predates headers".
	Bodies      bool `json:"bodies"`
	Headers     bool `json:"headers"`
	RedactQuery bool `json:"redactQuery"`
}

// reasonLimit keeps one failure message short enough to read in a tooltip.
const reasonLimit = 200

// failure reports the timeline event this line deserves, and whether it
// deserves one at all.
//
// Only a failure earns a mark. A console.log does not, and neither does a
// request that worked. The mark row is a map of a session, and a map that
// shows everything shows nothing.
func (ev LogEvent) failure() (Event, bool) {
	switch {
	case ev.T == LogError:
		return Event{T: "error", Reason: trimReason(ev.Text), TargetID: ev.TargetID}, true

	case (ev.T == LogConsole || ev.T == LogIssue) && ev.Level == "error":
		return Event{T: "error", Reason: trimReason(ev.Text), TargetID: ev.TargetID}, true

	case ev.T == LogNetFail:
		return Event{
			T: "netfail", URL: ev.URL, TargetID: ev.TargetID,
			Reason: trimReason(strings.TrimSpace(ev.Method + " " + ev.Text)),
		}, true

	case ev.T == LogRes && ev.Status >= 400:
		return Event{
			T: "netfail", URL: ev.URL, TargetID: ev.TargetID,
			Reason: trimReason(fmt.Sprintf("%s %d", ev.Method, ev.Status)),
		}, true
	}
	return Event{}, false
}

// failWindowMs is how close two identical failures have to be for the second
// one to raise a count instead of adding a mark.
const failWindowMs = 1000

// maxFailKeys caps the dedupe cache. It is a cache, not a record: a page that
// reports a thousand different errors must not grow it a thousand entries.
const maxFailKeys = 256

// failMark remembers a failure that is already on the timeline.
type failMark struct {
	idx  int   // where the event sits in the event slice
	last int64 // when it was last seen, on the recording clock
}

// failIndex collapses repeated failures into one mark that carries a count.
//
// A broken page reports the same error fifty times in a moment. Fifty marks in
// one place hide every other mark on the bar, so identical failures inside one
// second become one mark that says how many.
//
// The recorder feeds it one line at a time while it records, and a repair feeds
// it a whole journal afterwards. They share it so that a repaired manifest
// carries the marks a clean stop would have written.
type failIndex struct {
	seen map[string]*failMark
}

func newFailIndex() *failIndex { return &failIndex{seen: map[string]*failMark{}} }

// add returns events with the failure of ev folded into it, and events
// unchanged when ev is not a failure.
func (fx *failIndex) add(events []Event, ev LogEvent) []Event {
	e, ok := ev.failure()
	if !ok {
		return events
	}
	e.AtMs = ev.AtMs
	key := e.T + "\x00" + e.URL + "\x00" + e.Reason

	if m := fx.seen[key]; m != nil && ev.AtMs-m.last <= failWindowMs && m.idx < len(events) {
		events[m.idx].Count = atLeastOne(events[m.idx].Count) + 1
		m.last = ev.AtMs
		return events
	}
	if len(fx.seen) >= maxFailKeys {
		fx.seen = map[string]*failMark{}
	}
	e.Count = 1
	events = append(events, e)
	fx.seen[key] = &failMark{idx: len(events) - 1, last: ev.AtMs}
	return events
}

// IsFailure reports whether an event kind is one of the failures. The player
// draws those in red and ranks them first, and a list counts them.
func IsFailure(t string) bool { return t == "error" || t == "netfail" }

// manifestErrors is how many failures a finished recording holds.
//
// The count comes from the log block when there is one. A recording made
// before the log existed has no block, so its marks are all there is to count.
func manifestErrors(m *Manifest) int {
	if m.DevTools != nil {
		return m.DevTools.Errors
	}
	return countFailures(m.Events)
}

// countFailures totals the failures behind the marks. A mark that stands for
// fifty repeats counts fifty.
func countFailures(events []Event) int {
	n := 0
	for _, e := range events {
		if IsFailure(e.T) {
			n += atLeastOne(e.Count)
		}
	}
	return n
}

func trimReason(s string) string {
	s = strings.TrimSpace(strings.ReplaceAll(s, "\n", " "))
	if len(s) <= reasonLimit {
		return s
	}
	return s[:reasonLimit-1] + "…"
}

// logWriter appends the log journal of one recording.
//
// It is the same shape as the frame journal, and for the same reason: a
// recording that was killed still has everything up to the last line, so
// "atr record repair" can rebuild the marks from it.
type logWriter struct {
	buf  *bufio.Writer
	file *os.File
	max  int64

	lines   int
	bytes   int64
	dropped int
	full    bool
}

func newLogWriter(dir string, max int64) (*logWriter, error) {
	f, err := os.Create(filepath.Join(dir, devtoolsFile))
	if err != nil {
		return nil, fmt.Errorf("failed to create the log journal: %w", err)
	}
	switch {
	case max == NoMaxLog:
		max = 0 // zero here means no cap; the caller asked for that
	case max <= 0:
		max = DefaultMaxLog
	}
	return &logWriter{buf: bufio.NewWriter(f), file: f, max: max}, nil
}

// write appends one line. A line refused by the cap is counted, never an
// error: a full log must not end a recording.
func (w *logWriter) write(ev LogEvent) {
	if w == nil {
		return
	}
	// A drop line reports what somebody else refused, so it is counted even
	// when it is the line that does not fit.
	if ev.T == LogDrop {
		w.dropped += atLeastOne(ev.Count)
	}
	if w.full {
		if ev.T != LogDrop {
			w.dropped++
		}
		return
	}

	line, err := json.Marshal(ev)
	if err != nil {
		w.dropped++
		return
	}
	if w.max > 0 && w.bytes+int64(len(line))+1 > w.max {
		w.full = true
		if ev.T != LogDrop {
			w.dropped++
		}
		return
	}
	if _, err := w.buf.Write(append(line, '\n')); err != nil {
		w.dropped++
		return
	}
	w.lines++
	w.bytes += int64(len(line)) + 1
	// Flush often enough that a crash loses a handful of lines, and rarely
	// enough that the disk is not the bottleneck. The frame journal does the
	// same, for the same reason.
	if w.lines%20 == 0 {
		_ = w.buf.Flush()
	}
}

func (w *logWriter) stats() DevTools {
	if w == nil {
		return DevTools{}
	}
	return DevTools{Lines: w.lines, Bytes: w.bytes, Dropped: w.dropped}
}

func (w *logWriter) close() error {
	if w == nil {
		return nil
	}
	if err := w.buf.Flush(); err != nil {
		_ = w.file.Close()
		return fmt.Errorf("failed to flush the log journal: %w", err)
	}
	if err := w.file.Close(); err != nil {
		return fmt.Errorf("failed to close the log journal: %w", err)
	}
	return nil
}

func atLeastOne(n int) int {
	if n < 1 {
		return 1
	}
	return n
}

// recoverLog rebuilds the timeline marks and the log block of an interrupted
// recording from its log journal.
//
// The marks are the point. A session that was killed is the one somebody most
// wants to look at, and the failures in it are what they are looking for.
//
// It returns a nil block when the recording has no journal, because that is a
// recording from before the log existed and an empty block would claim the
// page said nothing. The capture flags come back false: only the manifest knew
// what the capture was allowed to keep, and that manifest was never written.
func recoverLog(dir string) ([]Event, *DevTools) {
	log, err := ReadLog(dir)
	if err != nil {
		return []Event{}, nil
	}

	events := []Event{}
	dt := DevTools{Lines: len(log)}
	fx := newFailIndex()
	for _, ev := range log {
		events = fx.add(events, ev)
		if ev.T == LogDrop {
			dt.Dropped += atLeastOne(ev.Count)
		}
	}
	if st, serr := os.Stat(filepath.Join(dir, devtoolsFile)); serr == nil {
		dt.Bytes = st.Size()
	}
	dt.Errors = countFailures(events)
	return events, &dt
}

// ReadLog reads the log journal of a recording.
//
// A truncated last line is expected after a crash, exactly as it is in the
// frame journal, so it is dropped rather than reported.
func ReadLog(dir string) ([]LogEvent, error) {
	f, err := os.Open(filepath.Join(dir, devtoolsFile))
	if err != nil {
		return nil, err
	}
	defer func() { _ = f.Close() }()

	var out []LogEvent
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 0, 64*1024), 1<<20)
	for sc.Scan() {
		var ev LogEvent
		if err := json.Unmarshal(sc.Bytes(), &ev); err != nil {
			continue
		}
		out = append(out, ev)
	}
	return out, nil
}

File: web/src/devtools.ts
// The model behind the DevTools dock.
//
// The wire and the journal carry the same rows, so the live dock and the
// player dock share everything here. A request arrives as two rows joined on
// reqId, and the dock shows one line per request, because that is what a person
// is looking at.

import type { LogEvent, LogLevel, RecEvent } from './protocol';

/** Tab is which part of the dock a row belongs to. */
export type Tab = 'console' | 'network' | 'issues';

/** ConsoleRow is one line in the Console tab. */
export interface ConsoleRow {
  key: string;
  atMs: number;
  /**
   * The wall clock the tap stamped. A live page has no recording to be
   * relative to, so this is the only time it can be shown at.
   */
  ts: number;
  level: LogLevel;
  text: string;
  stack?: string;
  url?: string;
  /** How many identical lines in a row this one stands for. */
  repeat: number;
}

/** NetRow is one request, with whatever came back joined onto it. */
export interface NetRow {
  key: string;
  atMs: number;
  /** See ConsoleRow.ts. */
  ts: number;
  method: string;
  url: string;
  kind: string;
  /** Undefined while the request is still open. */
  status?: number;
  bytes?: number;
  durMs?: number;
  /** Set when the request failed, was blocked, or was cancelled. */
  error?: string;
  /** True while nothing has come back, so the dock can say so. */
  pending: boolean;
}

export interface DevToolsModel {
  console: ConsoleRow[];
  network: NetRow[];
  issues: ConsoleRow[];
  /** Rows the rate cap or the size cap refused, so a gap says it is a gap. */
  dropped: number;
  /**
   * The origin most of the requests went to, which is the page's own host in
   * every session that is not a proxy. The dock hides this one and shows every
   * other, so a call to a third party cannot pass for a call to your server.
   *
   * It cannot be read from `location`: that is the ATR server, not the page.
   */
  origin: string;
  counts: { errors: number; warnings: number; failed: number };
}

export const EMPTY_MODEL: DevToolsModel = {
  console: [],
  network: [],
  issues: [],
  dropped: 0,
  origin: '',
  counts: { errors: 0, warnings: 0, failed: 0 },
};

/**
 * build turns the rows into the three tabs.
 *
 * It runs over the whole log every time it is called, which is what makes the
 * player dock and the live dock the same code. A live dock feeds it a ring of
 * two thousand rows, and a recording of an hour is tens of thousands: both are
 * a single pass over an array, and neither is worth an incremental index.
 */
export function build(rows: LogEvent[], limit = rows.length): DevToolsModel {
  const consoleRows: ConsoleRow[] = [];
  const issues: ConsoleRow[] = [];
  const network: NetRow[] = [];
  const byReq = new Map<string, NetRow>();
  let dropped = 0;

  for (let i = 0; i < limit; i += 1) {
    const ev = rows[i];
    switch (ev.t) {
      case 'console':
      case 'error':
        pushConsole(consoleRows, textRow(ev, i, ev.t === 'error' ? 'error' : ev.level ?? 'log'));
        break;

      case 'issue':
        pushConsole(issues, textRow(ev, i, ev.level ?? 'warning'));
        break;

      case 'tap':
        pushConsole(issues, textRow(ev, i, 'info'));
        break;

      case 'drop':
        dropped += ev.count && ev.count > 1 ? ev.count : 1;
        pushConsole(issues, textRow(ev, i, 'warning'));
        break;

      case 'req': {
        const row: NetRow = {
          key: `n${i}`,
          atMs: ev.atMs,
          ts: ev.ts,
          method: ev.method ?? '',
          url: ev.url ?? '',
          kind: ev.kind ?? '',
          pending: true,
        };
        network.push(row);
        if (ev.reqId) byReq.set(ev.reqId, row);
        break;
      }

      case 'res':
      case 'netfail': {
        // A res with no req before it is normal: a recording can start in the
        // middle of a page load. It becomes a row of its own rather than
        // vanishing.
        const row = ev.reqId ? byReq.get(ev.reqId) : undefined;
        const target = row ?? orphan(network, ev, i);
        target.pending = false;
        if (ev.method && !target.method) target.method = ev.method;
        if (ev.url && !target.url) target.url = ev.url;
        if (ev.kind && !target.kind) target.kind = ev.kind;
        if (ev.status) target.status = ev.status;
        if (ev.bytes) target.bytes = ev.bytes;
        if (ev.durMs) target.durMs = ev.durMs;
        if (ev.t === 'netfail') target.error = ev.text || 'failed';
        if (ev.reqId) byReq.delete(ev.reqId);
        break;
      }
    }
  }

  return {
    console: consoleRows,
    network,
    issues,
    dropped,
    origin: mainOrigin(network),
    counts: {
      errors: consoleRows.filter((r) => r.level === 'error').length,
      warnings:
        consoleRows.filter((r) => r.level === 'warning').length +
        issues.filter((r) => r.level === 'warning').length,
      failed: network.filter(isFailed).length,
    },
  };
}

/** mainOrigin is the origin the most requests went to. */
function mainOrigin(network: NetRow[]): string {
  const seen = new Map<string, number>();
  let best = '';
  let most = 0;
  for (const row of network) {
    let origin = '';
    try {
      origin = new URL(row.url).origin;
    } catch {
      continue;
    }
    const n = (seen.get(origin) ?? 0) + 1;
    seen.set(origin, n);
    if (n > most) {
      most = n;
      best = origin;
    }
  }
  return best;
}

/** isFailed is what the dock draws in red: a failure or a 4xx and up. */
export function isFailed(row: NetRow): boolean {
  return row.error !== undefined || (row.status !== undefined && row.status >= 400);
}

function textRow(ev: LogEvent, i: number, level: LogLevel): ConsoleRow {
  return {
    key: `c${i}`,
    atMs: ev.atMs,
    ts: ev.ts,
    level,
    text: ev.text ?? '',
    stack: ev.stack,
    url: ev.url,
    repeat: ev.count && ev.count > 1 ? ev.count : 1,
  };
}

/**
 * pushConsole collapses a line that repeats straight after itself. A render
 * loop logging the same warning sixty times a second is one line with a count
 * on it, exactly as a browser console shows it.
 */
function pushConsole(out: ConsoleRow[], row: ConsoleRow): void {
  const last = out[out.length - 1];
  if (last && last.level === row.level && last.text === row.text) {
    last.repeat += row.repeat;
    return;
  }
  out.push(row);
}

function orphan(network: NetRow[], ev: LogEvent, i: number): NetRow {
  const row: NetRow = {
    key: `n${i}`,
    atMs: ev.atMs,
    ts: ev.ts,
    method: ev.method ?? '',
    url: ev.url ?? '',
    kind: ev.kind ?? '',
    pending: false,
  };
  network.push(row);
  return row;
}

/** Identical warnings closer together than this become one mark. */
const WARN_WINDOW_MS = 1000;

/**
 * warnings turns the warning lines into timeline marks.
 *
 * The recorder does not write these. A warning is not a failure, and a bar with
 * a mark on every deprecation notice is a bar nobody can read, so they are
 * derived here and shown only while the chip is on. Deciding it at record time
 * would mean a chip that changed the next recording instead of this one.
 */
export function warnings(rows: LogEvent[]): RecEvent[] {
  const out: RecEvent[] = [];
  const seen = new Map<string, { mark: RecEvent; lastMs: number }>();
  for (const ev of rows) {
    if (ev.level !== 'warning') continue;
    if (ev.t !== 'console' && ev.t !== 'issue') continue;
    const text = (ev.text ?? '').replace(/\s+/g, ' ').trim().slice(0, 200);
    const prev = seen.get(text);
    // The window slides with each repeat, so a warning that fires steadily for
    // a minute is one mark, not sixty.
    if (prev && ev.atMs - prev.lastMs <= WARN_WINDOW_MS) {
      prev.mark.count = (prev.mark.count ?? 1) + 1;
      prev.lastMs = ev.atMs;
      continue;
    }
    const mark: RecEvent = { atMs: ev.atMs, t: 'warn', reason: text, count: 1 };
    out.push(mark);
    seen.set(text, { mark, lastMs: ev.atMs });
  }
  return out;
}

/** parse reads devtools.jsonl. A truncated last line is expected after a crash. */
export function parse(text: string): LogEvent[] {
  const out: LogEvent[] = [];
  for (const line of text.split('\n')) {
    if (!line) continue;
    try {
      out.push(JSON.parse(line) as LogEvent);
    } catch {
      // A half written last line is what an interrupted recording leaves.
    }
  }
  return out;
}

/**
 * countUpTo is how many rows the playhead has reached.
 *
 * It returns a count rather than a slice on purpose. A slice is a new array on
 * every call, so the model would be rebuilt ten times a second while the video
 * plays; a count only changes when a row actually crosses the playhead, which
 * is when the model actually changes.
 */
export function countUpTo(rows: LogEvent[], atMs: number): number {
  let lo = 0;
  let hi = rows.length;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (rows[mid].atMs <= atMs) lo = mid + 1;
    else hi = mid;
  }
  return lo;
}

/** matches is the filter box: a plain substring, case insensitive. */
export function matches(needle: string, ...fields: (string | undefined)[]): boolean {
  if (!needle) return true;
  const n = needle.toLowerCase();
  return fields.some((f) => f !== undefined && f.toLowerCase().includes(n));
}
